test/unit/meta/src/ApplyTupleTest.cpp

```diff
@@ -42,8 +42,10 @@ TEST_CASE("invoke", "[meta]")
     // invoke a free function
     REQUIRE(9 == alpaka::meta::invoke(abs_num, -9));
 
+#if !BOOST_COMP_PGI // nvc++: "a lambda is not allowed in an unevaluated expression"
     // invoke a lambda
     REQUIRE(42 == alpaka::meta::invoke([]() { return abs_num(-42); }));
+#endif
```

Comment:
Interesting observation. Lambdas in unevaluated expressions are a C++20 feature. Where is that lambda?

Reply:
My guess is, that this is a compiler bug and the lambda is not actually unevaluated.

nvc++ does not give comprehensive backtraces of instantiations like gcc would, so I do not know if the problem i somewhere deep inside catch2. Considering, that the very similar tests, with the `REQUIRE` and `meta::invoke` do compile, I think the lambda in question is the one which is the argument to `meta::invoke`.

Reply:
Checking the implementation of `alpaka::meta::invoke`, the lambda's type is actually used in unevaluated contexts (the `decltype` expressions in the return types of `invoke_impl`. Have you tried removing all the `decltype` expressions and replacing the `auto` return type by `decltype(auto)`?

Reply:
I think you are right, meaning the compiler may be correct in complaining, albeit a bit pedantic.

On the one hand the explicit return types look like just an artifact from the C++11 days. On the other, they do serve a purpose for SFINAE: When removing them, a call with a lambda leads to the last overload being used, which tries to call `<lambda_type>.<first arg as member pointer>` in the function body, which is an error. Having this expression on the `decltype` return type eliminates this variant from overload resolution.

So, you are saying, that alpaka already implicitly uses C++20? @psychocoderHPC what do you think about this?

Reply:
> On the one hand the explicit return types look like just an artifact from the C++11

In some places in alpaka explicit return types are required, else the compiler can not detect the return type.

> So, you are saying, that alpaka already implicitly uses C++20? @psychocoderHPC what do you think about this?

I am not aware that we use C++20, no compiler threw a warning or error.

Reply:
That `invoke` is using explicit return type is realy looking like a leftover from C++11. I would try what @bernhardmgruber  suggested.

https://github.com/alpaka-group/alpaka/blob/fe5c3f92388ee193784c5d41e13f49f01d09c074/include/alpaka/meta/ApplyTuple.hpp#L26-L56

Reply:
I already tried that, it breaks SFINAE, see above.

Reply:
We could replace the explicit return type by some `enable_if` but this would just move the lambda from one unevaluated context to another, as I understand it.

Reply:
I am back to thinking, that this is a compiler bug.

All compilers, including nvc++, seem to agree, that it is not allowed to have literal lambda (lambda-expression) in an unevaluated context, before c++20 . However, they also agree, that it is ok if you put any variable that is a lambda.
```c++
#include <utility>

template<class T, class... Args>
auto invoke(T&& t, Args... args)
	-> decltype(t(std::forward<Args...>(args...)))
	// OK
{
	return t(std::forward<Args...>(args...));
}

template<class... Args>
auto invokeL(Args... args)
	-> decltype([](int a){return a;}(std::forward<Args...>(args...)))
	// error: lambda-expression in unevaluated context
{
	return ([](int a){return a;}(std::forward<Args...>(args...)));
}

int b = 5;
auto lambda = [b = b](int a) mutable {return b += a;}; // making sure this thing has a closure

template<class... Args>
auto invokeLL(Args... args)
	-> decltype(lambda(std::forward<Args...>(args...)))
	// OK
{
	return (lambda(std::forward<Args...>(args...)));
}

int main()
{
	int b = 5;
	invoke([&b](int a){return a+b;}, 0);
	invokeL(0);
	invokeLL(0);
}
```

I do not see a lambda-expression in an unevaluated context in the in the involved Alpaka code.

Reply:
The behavior you describe here is fully conformant. You cannot define a lambda expression inside an unevaluated context, that is the `decltype(...)` expression here. The closure object created from the lambda expression, the `lambda` here, as well as any derived objects or types, can be used inside unevaluated contexts.

Reply:
> I do not see a lambda-expression in an unevaluated context in the in the involved Alpaka code.

You are right, there shouldn't be one. Still, you can remove all the `decltype` return types and replace `auto` by `decltype(auto)` to try to work around the bug.

Reply:
I cannot, because:

> On the one hand the explicit return types look like just an artifact from the C++11 days. On the other, they do serve a purpose for SFINAE: When removing them, a call with a lambda leads to the last overload being used, which tries to call `<lambda_type>.<first arg as member pointer>` in the function body, which is an error. Having this expression on the `decltype` return type eliminates this variant from overload resolution.

Reply:
Sorry, you are right! While you can remove it for some overloads, you cannot distinguish the member pointers (data member vs. function). Another call for C++17.